### ISSUE TITLE
BigGMr does not run in parallel

### ISSUE THREAD

**Issue by dpmcsuss** [jhu-graphstat/LSGMcode#2]
The way its constructed now if you want to specify a maxClusterSize rather than a specific number of clusters then you must use BigGMr. This has the somewhat tragic disadvantage that it does not run in parallel because it would have required nested parfor loops which are not allowed.

**Comment by dpmcsuss** [comment]
Apparently you can have nested `parfor` loops but we should still break out all clustering steps from the `parfor` loop.

**Comment by jtmatterer** [comment]
Take a look at 

[algorithms/kmeansBisect.m](https://github.com/jhu-graphstat/LSGMcode/blob/unequalGraphs/algorithms/kmeansBisect.m)

In the unequalGraphs branch. It runs kmeans on the data matrix and if any clusters exceed the max cluster size, it recurses on them. There are some issues with running it on large sets of vertices with very similar embeddings, bit it might be a good place to start.

**Comment by dpmcsuss** [comment]
I think the issue is that it won't work very well unless we re-embed the graphs. If we just cluster on the original points I would guess it would fail because those will already be somewhat tightly clustered as you note.

**Comment by dpmcsuss** [comment]
It would also be nice to write it in a way so that each recursion is run in parallel.

**Comment by jtmatterer** [comment]
I could do the implementation in parallel if we could create threads, but I don't think we can do that in marlab. Here's my design:

1) create a spectral clustering and alignment thread and start it. As it creates viable clusters, it adds them to a queue.
2) create a post clustering thread that handles reading in the subgraphs from the queue and generating a new thread to run sgm

For top-k results, some care will have to be taken so that there are no write race conditions, but I don't forsee there being any problems (other than having to reimplement everything in a different language).

**Comment by dpmcsuss** [comment]
I think that might be more than we need to start. I think its fine if fully run the clustering before we do any of the matching. For now getting the clustering running in parallel is a secondary priority though if its not too hard it would be nice to have.
